File: Problem056-Merge Intervals/Solution.py
"""
 @SEU(master);@ZZULI(bachelor)
 author:CuiXuan
 email:873059043@qq.com
 If you find any errors, please let me know.
 If you have any better solution, please let me know.

 Solution Using time 128ms
"""
# Definition for an interval.
# class Interval(object):
#     def __init__(self, s=0, e=0):
#         self.start = s
#         self.end = e

class Solution(object):
    def compare(self,a,b):
        if a.start != b.start:
            return a.start - b.start
        return a.end - b.end
    def merge(self, intervals):
        """
        :type intervals: List[Interval]
        :rtype: List[Interval]
        """
        returnColumn = []
        if len(intervals) == 0:return returnColumn
        intervals = sorted(intervals,cmp=self.compare)
        returnColumn.append(Interval(intervals[0].start,intervals[0].end))
        top=0
        for i in range(1,len(intervals)):
            if intervals[i].start <= returnColumn[top].end:
                returnColumn[top].end = max(intervals[i].end,returnColumn[top].end)
            else:
                returnColumn.append(Interval(intervals[i].start,intervals[i].end))
                top+=1
        return returnColumn
        

File: Problem056-Merge Intervals/Solution.java
/*
 * @SEU(master);@ZZULI(bachelor)
 * author:CuiXuan
 * email:873059043@qq.com
 * If you find any errors, please let me know.
 * If you have any better solution, please let me know.
 * 
 * Solution Using time 392ms
 */
/**
 * Definition for an interval.
 * public class Interval {
 *     int start;
 *     int end;
 *     Interval() { start = 0; end = 0; }
 *     Interval(int s, int e) { start = s; end = e; }
 * }
 */
public class Solution {
    public List<Interval> merge(List<Interval> intervals) {
        List<Interval> returnColumn = new ArrayList<Interval>();
        if(intervals.size() <= 0)return returnColumn;
        Collections.sort(intervals,new Comparator(){
            @Override
            public int compare(Object s,Object t){
                Interval a = (Interval)s;
                Interval b = (Interval)t;
                if(a.start != b.start){
                    return a.start - b.start;
                }else{
                    return a.end - b.end;
                }
            }
        });
        returnColumn.add(new Interval(intervals.get(0).start,intervals.get(0).end));
        int top=0;
        for(int i=1;i<intervals.size();i++){
            Interval val = intervals.get(i);
            Interval ret = returnColumn.get(top);
            if(val.start <= ret.end){
                ret.end = Math.max(ret.end,val.end);
                returnColumn.set(top,ret);
            }else{
                returnColumn.add(new Interval(val.start,val.end));
                top++;
            }
        }
        return returnColumn;
    }
}